### ISSUE TITLE
The beacon light is colliding with the glass

### ISSUE THREAD

**Issue by gilbertohasnofb** [c172p-team/c172p-detailed#682]
The beacon light is currently colliding with the glass:

![fg](https://cloud.githubusercontent.com/assets/5700795/12567948/9af5287a-c3c2-11e5-8b43-d66b393e9308.jpg)

See the discussion here for more info: https://github.com/Juanvvc/c172p-detailed/pull/678#issuecomment-174726479

**Comment by HHS81** [comment]
I gave the answer to solve this already 4 days ago here: https://github.com/Juanvvc/c172p-detailed/issues/633
Given in the tutorial in the last posting --> section Beacon. The trick is to move the emitting plane which presents the beacon light on the y-axis: 
![image](https://cloud.githubusercontent.com/assets/12649375/12593148/daa2d014-c46f-11e5-8504-358966796d2c.png)
The aircraft is seen here from the rear!

The xml it self does not need any modifications, the blend animation I used is just for this type of beacon (pulsating beacon).

**Comment by wlbragg** [comment]
>I gave the answer to solve this already 4 days ago here: #633

Apparently not, or I simply couldn't understand it.

I am trying part of the suggestion. I don't want to have to rely on a Python script for a different version of Blender than what I have installed, so I need to understand this manually without script.

So, I gathered you needed to offset the plane to at least the distance of the outer edge of the light object. I did that and I can see some changes but not enough. I have been able to get the light offset somewhat but not consistent with the view angle. The best I did was to get it from 3 out of 4 sides. The front view was still clipped.

One of my questions is the relationship of offset-m to the offset of the plane (object) and the axis-y or x?
In other words what is `<offset-m>5.61846</offset-m>`?

**Comment by wlbragg** [comment]
@HHS81 would it be possible for you to run your script on the c172p.blend in /dev and post the results for me, I would be very interested in seeing any differences from what we have?

**Comment by HHS81** [comment]
I understand that you don't want to run another version of Blender. 
The python script calculates the transational offset from the wanted point to the center (0,0,0).
It is more or less the distance from the point to the center. A bit geometry..... 

There is a method without a need for calculations: 
1.) create an emiting plane in blender at the center (0,0,0), and texture it
2.) save this as beaconlight.ac (or whatever)
3.) create a xml-file with following input:

```
<?xml version="1.0"?>
<PropertyList>
	<path>beaconlight.ac</path>

	<animation>
		<type>range</type><!-- LOD-->
		<min-m>0</min-m>
		<max-m>25000</max-m>
	</animation>

	<Beaconlightparams> <!-- size of the emitting plane depending of distance-->
		<light-near>0.4</light-near>
		<light-med>0.8</light-med>
		<light-far>10</light-far>
	</Beaconlightparams>

	
	
	<animation>
		<type>translate</type>
		<object-name>Beaconlight</object-name>
		<offset-m>0</offset-m>
		<axis>
			<x>0</x>
			<y>0</y>
			<z>0</z>
		</axis>
	</animation>

	<animation>
		<type>billboard</type>
		<object-name>Beaconlight</object-name>
		<spherical type="bool">true</spherical>
	</animation>

	<animation>
		<type>dist-scale</type><!-- distance for the plane size we had defined before-->
		<object-name>Beaconlight</object-name>
		<interpolation>
			<entry>
				<ind>0</ind>
				<dep alias="../../../../Beaconlightparams/light-near"/>
			</entry>
			<entry>
				<ind>500</ind>
				<dep alias="../../../../Beaconlightparams/light-med"/>
			</entry>
			<entry>
				<ind>16000</ind>
				<dep alias="../../../../Beaconlightparams/light-far"/>
			</entry>
		</interpolation>
	</animation>

</PropertyList>
```
4.) add the needed animation depending on voltage, switch etc . (the animation which makes the beacon blink) and call the beaconlight.xml from c172p.xml . Don't forget to delete the emitting plane in the c172p.ac.
`

**Comment by HHS81** [comment]
If it still don't work, I can create the beacon, but propably not before the weekend.

**Comment by wlbragg** [comment]
We already have all this in c172p.xml. I'm not opposed to trying to separate it out to test, but I guess I don't see why it should be any different of a result? It is the same as what we have only moved to an included file?
Do you know what `<offset-m>5.61846</offset-m>` is? What is that a measurement of?

**Comment by wlbragg** [comment]
>If it still don't work, I can create the beacon, but propably not before the weekend. 

I'm in no hurry as this may all be superseded by procedural texturing, but if you get a chance to use your script on the c172.blend, I would really like to see it!

Thank you for all your help and efforts.

**Comment by HHS81** [comment]
> Do you know what <offset-m>5.61846</offset-m> is? What is that a measurement of?

It is the translataional offset needed for the billboard animation. The translational offset is the offset from the wanted point to the center (0,0,0). Geometry, to make it short....

If the translational offset is calculated wrong, the light appears anywhere else, but not there where you want it. 
While the procedural texturing offered by Thorsten R. sounds nice, it is always good to have a fallback for non-shader users.

**Comment by onox** [comment]
@wlbragg Just in case you're using that XML from @HHS81 above, please instead of `<dep alias="../../../../Beaconlightparams/light-near"/>` use something like `<dep alias="/params/beacon-light/light-near"/>`.

**Comment by wlbragg** [comment]
I haven't changed any of the code, we were already using the same method as far as I was aware of.

**Comment by wlbragg** [comment]
Looks like the proceedural light code is in place. I am a bit under the weather today and may not start on this until I feel a little more like the "living".

**Comment by gilbertohasnofb** [comment]
No rush about it and get better soon :smile_cat:

**Comment by gilbertohasnofb** [comment]
@wlbragg I just checked and this is still an issue even after we implemented the procedural lights. Also, the problem can be seen from any angle. See:

![fg-0](https://cloud.githubusercontent.com/assets/5700795/17461766/fe5c2bf2-5c97-11e6-906c-219e2c50cb62.jpg)

![fg-1](https://cloud.githubusercontent.com/assets/5700795/17461765/fe5b214e-5c97-11e6-8ef9-8caf9fdd2cc8.jpg)

**Comment by onox** [comment]
AFAIK we only used the ALS procedural lights for the nav lights. Beacon and landing/taxi should still be non-shader lights.

**Comment by wlbragg** [comment]
I was not skilled enough to fix this.

**Comment by gilbertohasnofb** [comment]
No problems. But is the beacon light procedurally generated or does it use a non-shader light?

**Comment by gilbertohasnofb** [comment]
The c182s seems to use a procedural light for the beacon, and it manages to avoid this collision issue we are having. Maybe @HHS81 can give us an opinion (or two)? :stuck_out_tongue_closed_eyes:

**Comment by wlbragg** [comment]
>does it use a non-shader light?

Yes

Absolutely, lets look at the c182 and see how it's done.

**Comment by legoboyvdlp** [comment]
Just looking through old issues, I definitely saw this mentioned recently, was this fixed in the remodelling efforts?

**Comment by wlbragg** [comment]
No, it was slightly adjusted but not fixed.

**Comment by gilbertohasnofb** [comment]
@wlbragg what about hiding the lamp object while the flash is on? Is that too much of a hack?

**Comment by wlbragg** [comment]
>is that too much of a hack?

I had the same thought but figured it wouldn't look right. I think you'd notice the light disappearing during a flash. I worked with it for over an hour yesterday trying to get it to look right, but no luck .The c182s doesn't do what ours is doing and I can't figure out why?

**Comment by legoboyvdlp** [comment]
How about making the light itself transparent, but with a red tinted glass...?

**Comment by onox** [comment]
Isn't it possible to move the quad in the .ac a little bit to the front, so that the quad is always positioned in front of the object it is positioned at?

**Comment by legoboyvdlp** [comment]
What about looking at the beacon from the rear?

**Comment by onox** [comment]
The quad's orientation is always such that it faces the camera.

**Comment by wlbragg** [comment]
>The quad's orientation is always such that it faces the camera.

I looked at the c182s in Blender and I though ours was identical, I'll look again with this bit of information and see if I missed something. @onox do you know what the offset is for in the beacon code?
~~~
<animation>
        <type>translate</type>
        <object-name>BeaconOffX</object-name>
        <offset-m>5.618312</offset-m>
        <axis>
            <!-- MOD: Position beacon light at MEDIAN point of the object -->
            <x> 5.49464</x>
            <y>-0.00000</y>
            <z> 1.11300</z>
            <!-- MODEND: Position beacon light at MEDIAN point of the object -->
        </axis>
    </animation>
~~~

**Comment by onox** [comment]
It's used like this: translation = offset-m + factor * (property | expression | 1)

**Comment by onox** [comment]
But the offset has to happen in the direction towards the camera, so it should be done in the .ac file or in the vertex shader.

**Comment by HHS81** [comment]
@onox 
@wlbragg 
As can remember it is done in the .ac

**Comment by wlbragg** [comment]
>It's used like this: translation = offset-m + factor * (property | expression | 1)

In the case of the Beacon...
What is the offset for? 
Why is it used?
What does it do?

**Comment by wlbragg** [comment]
OK, it's fixed. Don't ask me how though because I really don't have a solid answer. I have a feeling though the coordinate directions are mixed up VS what we use for other animations and that is why it never made any sense to me looking at examples.

I hope this is good enough, as I am done with it.

**Comment by legoboyvdlp** [comment]
Issue should be closed if it is fixed.

**Comment by gilbertohasnofb** [comment]
@legoboyvdlp you are right, but in this particular case the PR with the fix hasn't been merged yet so the issue is not fixed on the master just yet :wink:
